### PR TITLE
journal: enforce strict ordering only when writing journal files from journald, but not from journal-remote and similar

### PR DESCRIPTION
**Pull request by poettering**
A follow-up for 1d8d483f59ffa62974772fb58a8ef4abe88550ec to condition (and extend) the logic a bit.

### CODE REVIEW

File: src/journal/journald-server.c [hunk 1]
```diff
@@ -844,8 +848,16 @@ static bool shall_try_append_again(JournalFile *f, int r) {
...
-                log_ratelimit_warning(JOURNAL_LOG_RATELIMIT, "%s: Time jumped backwards relative to last journal entry, rotating.", f->path);
+        case -EREMCHG:         /* Wallclock time (CLOCK_REALTIME) jumped backwards relative to last journal entry */
+                log_ratelimit_warning(JOURNAL_LOG_RATELIMIT, "%s: Realtime clock jumped backwards relative to last journal entry, rotating.", f->path);
+                return true;
+
+        case -EREMOTE:         /* Boot ID different from the one of the last entry */
+                log_ratelimit_warning(JOURNAL_LOG_RATELIMIT, "%s: Boot ID changed since last record, rotating.", f->path);
+                return true;
+
+        case -ENOTNAM:         /* Monotonic time (CLOCK_MONOTONIC) jumped backwards relative to last journal entry */
```
Comment (keszybz) [inline src/journal/journald-server.c:885]: Let's just hope this errno never "leaks". "Not a XENIX named type file" could be a bit confusing.

**Review by keszybz — COMMENTED** [review]
The relaxation of requirements for journal-remote and rotation seems like something that we very much should do.

The additional checks on monotonic timestamps is a separate topic. But it's essentialy 10 lines of very simple code… so it should be OK, even after -rc1.

LGTM.

**Comment by keszybz** [comment]
I restarted jammy-*.

**Comment by poettering** [comment]
@DaanDeMeyer could you bless this?

**Review by DaanDeMeyer — APPROVED** [review]

File: src/journal/journald-server.c (hunk 1 above)
Comment (mrc0mmand) [reply src/journal/journald-server.c:885]: *Narrator: the errno leaked and caused confusion.*

**Comment by smopucilowski** [comment]
@keszybz I did indeed hit the puzzling errno mentioned above:

```
[ 2988.374664] systemd-journald[328]: /var/log/journal/bd3be28df719f3499cdef198566abc2e/user-1000.journal: Montonic clock jumped backwards relative to last journal entry, rotating.
[ 2988.374673] systemd-journald[328]: Failed to write entry to /var/log/journal/bd3be28df719f3499cdef198566abc2e/user-1000.journal (28 items, 781 bytes), rotating before retrying: Not a XENIX named type file
```

**Comment by rokezu** [comment]
@keszybz I also hit the errno, no idea what is it about, confusing indeed, and a a typo with the "Montonic" word as well.  Seems benign, and in this case, it is a system moved/rsynced to a new drive, so maybe file attributes were lost?

```
[   22.386669] systemd-journald[988]: /var/log/journal/4af61aba1b7d49c4987f7e2aa5e2ac4d/user-1000.journal: Montonic clock jumped backwards relative to last journal entry, rotating.
[   22.386683] systemd-journald[988]: Failed to write entry to /var/log/journal/4af61aba1b7d49c4987f7e2aa5e2ac4d/user-1000.journal (32 items, 1139 bytes), rotating before retrying: Not a XENIX named type file
```

**Comment by ManuLinares** [comment]
What is this "Montonic clock jumped backwards relative to last journal entry, rotating." on my journal?

I can't find anything but this pull

**Comment by semeion** [comment]
> What is this "Montonic clock jumped backwards relative to last journal entry, rotating." on my journal?
> 
> I can't find anything but this pull

Yeah, same message happening here in 3 different machines using Arch Linux. I have no idea the reason for that weird message.

**Comment by keszybz** [comment]
Guys, the confusing message was adjusted/fixed in #26605, which is not in any released version yet. Please don't report the **old** message, we know it's confusing, that's why we changed it. Please instead check if the issue is fixed with the patch (systemd-254 or in the backport branches).

File: src/libsystemd/sd-journal/journal-file.c [hunk 2]
```diff
@@ -2092,11 +2096,37 @@ static int journal_file_append_entry_internal(
...
+                 * Typically, if any of the errors generated here are seen journald will just rotate the
+                 * journal files and start anew. */
+
+                if (ts->realtime < le64toh(f->header->tail_entry_realtime))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EREMCHG),
+                                               "Realtime timestamp %" PRIu64 " smaller than previous realtime "
+                                               "timestamp %" PRIu64 ", refusing entry.",
+                                               ts->realtime, le64toh(f->header->tail_entry_realtime));
+
+                if (!sd_id128_is_null(f->header->boot_id) && boot_id) {
```
Comment (msizanoen1) [inline src/libsystemd/sd-journal/journal-file.c:2117]: @poettering: A bit late, but do we really need to ensure consistency of boot ID and monotonic timestamp (both of which resets across a system reboot) to ensure deterministic bisection considering that we already enforced strict order for the real time clock? This is causing potentially unnecessary rotation every reboot, possibly wasting storage space for no benefit.

**Pull request merged**